tracer.go

```diff
@@ -29,6 +29,41 @@ type Options struct {
 	// attaching external code to trace events. See NetTraceIntegrator for a
 	// practical example, and event.go for the list of possible events.
 	NewSpanEventListener func() func(SpanEvent)
+	// DebugAssertSingleGoroutine internally records the ID of the goroutine
+	// creating each Span and verifies that no operation is carried out on
+	// it on a different goroutine.
+	// Provided strictly for development purposes.
+	// Passing Spans between goroutine makes it notoriously difficult to avoid
+	// use-after-free errors, especially in the presence of timeouts.
```

Comment:
nit: `use-after-Finish() errors`

non-nit: I agree that it's an anti-pattern for two goroutines to make use of a Span when it's unclear when/where Finish() might be called. That said, for general-purpose producer/consumer `chan`s it is not hard to imagine a Span moving between goroutines, perhaps via a `Context` object/etc, without any cause for alarm. So the salient point is that passing a Span between goroutines is fine if.f. there is never any doubt about which goroutine will eventually call Finish(), and whether any other goroutines might have access to the Span whenever that happens. If the above makes sense, perhaps the comment could be clearer about the parameters of dangerous/bad situation?

BTW, in C++ Dapper, all of Finish() was handled via refcounting (dating back to Mike Burrows' initial experiments with automatic tracing across google3 applications)... i.e., the equivalent of Finish() was invoked implicitly whenever the last reference to the "Span" (actually "TraceContext" at google) was dropped. Unfortunately there were plenty of "runaway Spans" where a "permanent" ref would get picked up by the google3 closure library and screw up various buffers and statistics as a result. On the plus side, programmers didn't need to directly worry about the sort of thing you're documenting here. In retrospect: a net loss. Much happier with explicit calls to Finish() here.